### PR TITLE
Support using colon to omit a dimension in reshape

### LINKED ISSUE #16790: Allow : in reshape
I find myself writing a lot of this:

``` jl
A = reshape(A, N, div(length(A), N))
```

Instead, I'd like to write 

``` jl
A = reshape(A, N, :)
```

### PR DESCRIPTION
**Pull request by mbauman**
Closes #16790.

### CODE REVIEW

File: base/reshapedarray.jl [hunk 1]
```diff
@@ -36,10 +36,78 @@ start(R::ReshapedArrayIterator) = start(R.iter)
...
 length(R::ReshapedArrayIterator) = length(R.iter)
 
+"""
+    reshape(A, dims...)
+    reshape(A, dims)
+
+Return an array with the same data as the given array, but with different dimensions.
+
+The new dimensions may be specified either as a list of arguments or as a shape
+tuple. Up to one dimension may be specified with a `:`, in which case its
```
Comment (tkelman) [inline base/reshapedarray.jl]: at most one?

Comment (tkelman) [inline base/reshapedarray.jl]: (I'd change it in the error message too)

**Comment by tkelman** [comment]
This would be useful, did anyone have opinions on this?

**Comment by tkelman** [comment]
just in case? @nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by tknopp** [comment]
This is a very useful feature. So it would be great to have this for 0.6. Should be added to NEWS once merged.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19919#issuecomment-273027255) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/990e4b30980d183af22bbfe7018f33618c5a5763/7e132e6_vs_a1f232e/report.md). cc @jrevels

**Comment by tkelman** [comment]
I think those are just noise, but for good measure @nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Review by tkelman — APPROVED** [review]

**Comment by pabloferz** [comment]
Those noisy results are the ones that #20025 will fix.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19919#issuecomment-273357923) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/25fda874a385d46fdb0109aa2b2d4f9d8fc40075/26e1969_vs_455b52f/report.md). cc @jrevels

**Comment by mbauman** [comment]
NEWS.md added!

**Pull request merged**